docs/specs/alcproxies.md

```diff
@@ -0,0 +1,251 @@
+
+In the past, a common use of AppDomain was to load an assembly into a separate execution environment, where developers were able to load objects and use them as if they were in the original domain. 
+
+
+## Comparison
+
+### AppDomain
+For AppDomains, developers could load objects and execute methods like so (Shortened version from [here](https://docs.microsoft.com/en-us/dotnet/api/system.appdomain?view=netframework-4.8 )):
+
+
+```C#
+public class SampleType : MarshalByRefObject
+{
+    // Call this method via a proxy.
+    public void SomeMethod()
+    {
+        Console.WriteLine("Hello!");
+    }
+}
+```
+
+...and then within the main application...
+
+
+```C#
+...
+
+AppDomainSetup ads = new AppDomainSetup();
+ads.ApplicationBase = AppDomain.CurrentDomain.BaseDirectory;
+
+...
+
+AppDomain ad2 = AppDomain.CreateDomain("AD #2", null, ads);
+
+SampleType mbrt = 
+            (SampleType) ad2.CreateInstanceAndUnwrap(
+                exeAssembly, 
+                typeof(SampleType).FullName
+            );
+			
+mbrt.SomeMethod();
+
+...
+```
+
+This allows for programs and classes to be loaded with an extra boundary of isolation, to help prevent things like unstable code bringing down a process. This was removed due to the support for the implementation (`TransparentProxy`) being fairly expensive.
+
+### AssemblyLoadContext
+Currently, there is no easy way to load an object from an ALC and call methods normally on it. However, you can use reflection like so:
+```C#
+AssemblyLoadContext alc = new AssemblyLoadContext(assemblyPath, true); //Create the ALC
+Assembly aa = alc.LoadFromAssemblyPath(assemblyPath); //Load the assembly into the ALC
+Type classType = aa.GetType("AssemblyName.DesiredType");
+
+// For instantiating below, you can also use Activator.GetInstance(classType) if you know there is a default constructor that takes no arguments
+object obj = (classType.GetConstructor(Type.EmptyTypes).Invoke(new object[] { }));
+
+obj.GetType().GetMethod("SomeMethod").Invoke(obj, null);
+```
+The problem here is that due to Type Isolation of ALCs, you cannot directly cast the instantiated object to its original type or an inherited type, as the CLR believes them to be different types from what's loaded from the ALC. 
+
+There are two ways to use the instance from the ALC:
+* Use reflection to invoke calls of the objects type, gotten directly using object.GetType()
+* Use a shared assembly between the user and target assembly that holds a class/interface that the type you want inherits from, and using said shared type instead of the type you want directly (See [this example](https://github.com/dotnet/samples/tree/master/core/extensions/AppWithPlugin ) to see it working).
+* Use `DispatchProxy` to create a proxy within the user ALC, pointing to an object in the target ALC. This also requires a shared assembly.
```

Comment:
Another option is to use the [Type Equivalence](https://docs.microsoft.com/dotnet/framework/interop/type-equivalence-and-embedded-interop-types) feature through COM. The attribute is [`TypeIdentifier`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.typeidentifierattribute). This was a feature designed for COM but it is an approach that could be reused. There are many issues with this though.